digicampipe/image/utils.py

```diff
@@ -0,0 +1,416 @@
+        center_peaks.append(np.array((xcenter + max_pos_x - 20, ycenter + max_pos_y - 20)))
+        for ix, x in enumerate(range(max_pos_x - 20, max_pos_x + 20)):
+            for iy, y in enumerate(range(max_pos_y - 20, max_pos_y + 20)):
+                auto_correlation[y, x] -= gaussian_function(ix, iy)
+    center_peaks = np.array(center_peaks)
+    """
+    auto_correlation[auto_correlation < 0] = 0
+    mean, median, std = sigma_clipped_stats(auto_correlation, sigma=3.0, iters=5)
+    baseline_sub = auto_correlation - median
+    baseline_sub[baseline_sub < 0] = 0
+    daofind = DAOStarFinder(fwhm=3.0, threshold=std)
+    sources = daofind(baseline_sub)
+    center_peaks = np.array([sources['xcentroid'], sources['ycentroid']]).transpose()
+    order = np.argsort(sources['mag'])
+    center_peaks = center_peaks[order[0:100], :]
+    ks = center_peaks[1:, :] - center_peaks[0, :]
+    ks_complex = ks[:, 0] + 1j * ks[:, 1]
+    angles = np.angle(ks_complex)
+    lengths = np.abs(ks_complex)
+    short_to_long = np.argsort(lengths)
+    ks_base = np.empty((2, 2))
+    ks_base[0, :] = ks[short_to_long[0], :]
+    angles = angles - np.angle(ks_base[0, 0]+1j*(ks_base[0, 1]))
+    factors = np.abs(ks_complex/(ks_base[0, 0]+1j*(ks_base[0, 1])))
+    is_collinear = (np.abs(np.mod(angles+np.pi, 2*np.pi)-np.pi) < 1*np.pi/180)  # less than 1 deg from 0
+    is_anti_collinear = np.abs(np.abs(angles)-np.pi) < 1*np.pi/180  # less than 1 deg from pi
+    factors[is_anti_collinear] *= -1
+    is_multiple = np.abs(np.mod(factors + 0.5, 1) - 0.5) < 1e-2
+    is_used = is_multiple & (is_collinear | is_anti_collinear)
+    rounded_factors = np.round(factors[is_used]).reshape(-1, 1)
+    ks1s_rescaled = 1 / rounded_factors * ks[is_used, :]
+    ks_base[0, :] = np.mean(ks1s_rescaled, axis=0)
+    for i in short_to_long[1:]:
+        if is_collinear[i] or is_anti_collinear[i]:
+            continue  # do not take collinear vector as base
+        ks_base[1, :] = ks[i]
+        angles = np.angle(ks_complex / (ks_base[1, 0] + 1j * (ks_base[1, 1])))
+        factors = np.abs(ks_complex / (ks_base[1, 0] + 1j * (ks_base[1, 1])))
+        is_collinear = (np.abs(np.mod(angles + np.pi, 2 * np.pi) - np.pi) < 1 * np.pi / 180)  # less than 1 deg from 0
+        is_anti_collinear = np.abs(np.abs(angles) - np.pi) < 1 * np.pi / 180  # less than 1 deg from pi
+        factors[is_anti_collinear] *= -1
+        is_multiple = np.abs(np.mod(factors + 0.5, 1) - 0.5) < 1e-2
+        is_used = is_multiple & (is_collinear | is_anti_collinear)
+        rounded_factors = np.round(factors[is_used]).reshape(-1, 1)
+        ks2s_rescaled = 1 / rounded_factors * ks[is_used, :]
+        ks_base[1, :] = np.mean(ks2s_rescaled, axis=0)
+        break
+    ks_base[ks_base[:, 0] < 0, :] *= -1
+    return ks_base, auto_correlation_saved, center_peaks
+
+
+def get_image_hexagonalicity(image, rotations=(60, 300)):
+    # init = clock()
```

Comment:
such comments should not end up in production code. It is possible to benchmark fuctions in pytests